cdm/core/src/main/java/ucar/nc2/filter/ConvertMissing.java

```diff
@@ -111,42 +112,62 @@ public static ConvertMissing createFromVariable(VariableDS var) {
             // TODO add logger
           }
         }
-        hasMissingValue = true;
       } else { // not a string
         missingValue = new double[missingValueAtt.getLength()];
         DataType missingType = FilterHelpers.getAttributeDataType(missingValueAtt, signedness);
         for (int i = 0; i < missingValue.length; i++) {
+          double val = var.convertUnsigned(missingValueAtt.getNumericValue(i), missingType).doubleValue();
           missingValue[i] = var.convertUnsigned(missingValueAtt.getNumericValue(i), missingType).doubleValue();
           missingValue[i] = var.applyScaleOffset(missingValue[i]);
         }
-
-        for (double mv : missingValue) {
-          if (!Double.isNaN(mv)) {
-            hasMissingValue = true; // dont need to do anything if it's already a NaN
-            break;
-          }
-        }
       }
     }
     return new ConvertMissing(var.fillValueIsMissing(), var.invalidDataIsMissing(), var.missingDataIsMissing(),
-        hasValidMin, hasValidMax, validMin, validMax, hasFillValue, fillValue, hasMissingValue, missingValue);
+        hasValidMin, hasValidMax, validMin, validMax, hasFillValue, fillValue, missingValue);
   }
 
 
   public ConvertMissing(boolean fillValueIsMissing, boolean invalidDataIsMissing, boolean missingDataIsMissing,
       boolean hasValidMin, boolean hasValidMax, double validMin, double validMax, boolean hasFillValue,
-      double fillValue, boolean hasMissingValue, double[] missingValue) {
+      double fillValue, double[] missingValue) {
     this.fillValueIsMissing = fillValueIsMissing;
     this.invalidDataIsMissing = invalidDataIsMissing;
     this.missingDataIsMissing = missingDataIsMissing;
     this.hasValidMin = hasValidMin;
     this.hasValidMax = hasValidMax;
     this.validMin = validMin;
+    this.fuzzyValidMin = validMin - Misc.defaultMaxRelativeDiffFloat;
     this.validMax = validMax;
+    this.fuzzyValidMax = validMax + Misc.defaultMaxRelativeDiffFloat;
     this.hasFillValue = hasFillValue;
     this.fillValue = fillValue;
-    this.hasMissingValue = hasMissingValue;
     this.missingValue = missingValue;
+    this.hasMissingValue = false;
+    // clean up missing values: remove NaNs, fill values, and values outside valid range
+    if (this.missingDataIsMissing && this.missingValue != null) {
+      List<Double> missing = new ArrayList();
+      for (double mv : this.missingValue) {
+        if (mv == Double.NaN) {
```

Comment:
I think this should be
```suggestion
        if (Double.isNaN(mv)) {
```